File: Zamba/ZambaWeb2013/zamba2010/Zamba.WebWFMonitor/UserControls/Login.ascx.cs
using System;
using System.Data;
using System.Configuration;
using System.Web;
using System.Web.Security;
using System.Web.UI;
using System.Web.UI.WebControls;
using System.Web.UI.WebControls.WebParts;
using System.Web.UI.HtmlControls;
using Zamba.Servers;
using Zamba.Users.Factory;
using Zamba.Core;

public partial class Login : System.Web.UI.UserControl
{
    public delegate void SuccesfullLogIn();
    public event SuccesfullLogIn LoggedIn;
    private Int32 _loginAttempts = 0;
    private Boolean _remeberUser = false;
    private User _user = null;

    private void Login_Fail(String Usuario)
    {
        _loginAttempts += 1;
        if (_loginAttempts == 3)
        {
            String IsUserLocked;
            IConnection Con = Zamba.Servers.Server.get_Con(false, false, false);
            Con.Open();
            IsUserLocked = (String)Con.ExecuteScalar(CommandType.Text, "Select lockuser from ZSecOption");
            Con.Close();
            Con.dispose();
            if (Boolean.Parse(IsUserLocked))
                SecurityOptions.LockUserPassword(Int32.Parse(this.ID), Usuario);

            RightFactory.SaveAction(GetUserID(Usuario), IUser.ObjectTypes.Users, IUser.RightsType.InicioFallidoDeSesion, "PC:" + Environment.MachineName + ",Usuario_Windows:" + Environment.UserName, GetUserID(Usuario));
            Response.Redirect("about:blank");
        }
    }

    private static Int32 GetUserID(String name)
    {
        String Query = "SELECT id FROM usrtable WHERE name='" + name + "'";
        IConnection Con = Zamba.Servers.Server.get_Con(false, false, false);
        Con.Open();
        Int32 UserId = Int32.Parse((Con.ExecuteScalar(CommandType.Text, Query).ToString()));
        Con.Close();
        Con.dispose();

        return UserId;
    }

    private void Validar(String nombreUsuario, String password)
    {
        try
        {
           _user = RightFactory.ValidateLogIn(nombreUsuario, password);
            if (_user == null)
                Login_Fail(nombreUsuario);

            if (_user.ID  > 0)
            {
                if (!SecurityOptions.ClaveVencida((Int32)_user.ID))
                    Login_Ok(nombreUsuario);
                else
                {
                    pnlLog.Visible = false;
                    pnlCambioPassword.Visible = true;
                }
            }
            else
            {
                if (_user.ID == -1)
                    FailureText.Text = "El Usuario se encuentra bloqueado";
                else
                    FailureText.Text = "El Usuario o la clave ingresadas son incorréctos";

                Login_Fail(nombreUsuario);
            }
        }
        catch (System.Threading.ThreadAbortException ex)
        {
            ZClass.RaiseError(ex);
        }
        catch (Exception ex)
        {
            ZClass.RaiseError(ex);
            FailureText.Text = ex.Message;
            Login_Fail(nombreUsuario);
        }
    }

    private void Login_Ok(String usuario)
    {
        try
        {
            String WinUser = Environment.UserName;
            String WinPc = Environment.MachineName;
            _remeberUser = cbRememberMe.Checked;
            if (_remeberUser)
            {
                if (RightFactory.GetUserRights(IUser.ObjectTypes.ModuleWorkFlow, IUser.RightsType.Use, -1) == false)
                    _remeberUser = false;
            }

            RightFactory.CurrentUser.WFLic = _remeberUser;
            RightFactory.CurrentUser.ConnectionId = Ucm.NewConnection((Int32)RightFactory.CurrentUser.ID, WinUser, WinPc, UserPreferences.TimeOut, _remeberUser);
            Ucm.ConectionTime = DateTime.Now;

            _remeberUser = cbRememberMe.Checked;
            if (_remeberUser)
            {
                if (RightFactory.GetUserRights(IUser.ObjectTypes.ModuleWorkFlow, IUser.RightsType.Use, -1) == false)
                    _remeberUser = false;
            }
            RightFactory.CurrentUser.WFLic = _remeberUser;
            RightFactory.CurrentUser.ConnectionId = Ucm.NewConnection((Int32)RightFactory.CurrentUser.ID, _user.Name, WinPc, UserPreferences.TimeOut, _remeberUser);
            Ucm.ConectionTime = DateTime.Now;

            LoggedIn();
        }
        catch (Exception ex)
        {
            ZClass.RaiseError(ex);
            this.FailureText.Text = "Maximo de Licencias conectadas, contáctese con su proveedor para adquirir nuevas licencias.";
            RightFactory.CurrentUser.ConnectionId = 0;
        }
    }
        
    protected void LoginButton_Click(object sender, System.EventArgs e)
    {
        Validar(this.tbNombreUsuario.Text, this.tbContraseña.Text);
    }
    
    protected void btCambiarPassword_Click(object sender, EventArgs e)
    {
        if (SecurityOptions.IsValidPassword((Int32)_user.ID, tbNombreUsuario.Text, tbContraseña2.Text.Trim(), true))
        {
            _user.Password = tbContraseña2.Text.Trim();
            try
            {
                UserFactory.Update(this._user);
                this.pnlCambioPassword.Visible = false;
                this.pnlLog.Visible = true;
            }
            catch (Exception ex)
            {
                ZClass.RaiseError(ex);
                Login_Fail(tbNombreUsuario.Text);
            }
        }
    }
}

File: Zamba/zambawindows2013/Zamba2010/Zamba.WebWFMonitor/UserControls/Login.ascx.vb
Imports Zamba.Data
Imports Zamba.Core

Partial Class Login
    Inherits System.Web.UI.UserControl

    Public Event Logueado()
    Dim Count As Integer
    Dim User As String
    Dim Pass As String
    Dim WF As Boolean
    Dim MyUser As User

    Private Sub Login_Fail(ByVal Usuario As String)
        Count += 1
        If Count = 3 Then
            If Zamba.Servers.Server.Con.ExecuteScalar(System.Data.CommandType.Text, "Select lockuser from ZSecOption") = "TRUE" Then
                SecurityOptions.LockUserPassword(Me.ID, Usuario)
            End If
            Zamba.Users.Factory.RightFactory.SaveAction(GetUserID(Usuario), IUser.ObjectTypes.Users, IUser.RightsType.InicioFallidoDeSesion, "PC:" & Environment.MachineName & ",Usuario_Windows:" & Environment.UserName, GetUserID(Usuario))
            Response.Redirect("about:blank")
        End If
    End Sub


    Private Shared Function GetUserID(ByVal name As String) As Int32
        Try
            Dim sql As String = "Select id from usrtable where name='" & name & "'"
            Return Zamba.Servers.Server.Con.ExecuteScalar(System.Data.CommandType.Text, sql)
        Catch
            Return 0
        End Try
	End Function


    Private Sub Login_Ok(ByVal Usuario As String)
        Try

            Dim WinUser, WinPc As String
            Try
                WinUser = Environment.UserName
            Catch ex As Exception
                RaiseError(ex)
                WinUser = "Desconocido"
            End Try
            Try
                WinPc = Environment.MachineName
            Catch ex As Exception
                RaiseError(ex)
                WinPc = "Desconocida"
            End Try
            'Validar licencia para WorkFlow, parametro opcional WF
            Me.WF = Me.RememberMe.Checked
            If Me.WF = True Then
                If Zamba.Users.Factory.RightFactory.GetUserRights(IUser.ObjectTypes.ModuleWorkFlow, IUser.RightsType.Use) = False Then
                    'No tiene permiso para acceder a WF
                    WF = False
                End If
            End If
            Zamba.Users.Factory.RightFactory.CurrentUser.WFLic = WF
            Zamba.Users.Factory.RightFactory.CurrentUser.ConnectionId = Ucm.NewConnection(Zamba.Users.Factory.RightFactory.CurrentUser.ID, WinUser, WinPc, UserPreferences.TimeOut, WF)
            Ucm.ConectionTime = Now.ToString("hh:mm:ss")
        Catch ex As Exception
            RaiseError(ex)
            Me.FailureText.Text = "Maximo de Licencias conectadas, contáctese con su proveedor para adquirir nuevas licencias."
            Zamba.Users.Factory.RightFactory.CurrentUser.ConnectionId = 0
        End Try
        Try
            UserPreferences.UserId = Zamba.Users.Factory.RightFactory.CurrentUser.ID
        Catch
        End Try
        If Zamba.Users.Factory.RightFactory.CurrentUser.ConnectionId <> 0 Then
            Session("Usuario") = Usuario
            Dim L As Generic.List(Of Usuario) = Application("Usuarios")
            If IsNothing(L) Then L = New Generic.List(Of Usuario)
            L.Add(New Usuario(Usuario))
            Application("Usuarios") = L

            RaiseEvent Logueado()
        Else
            Me.FailureText.Text = "Límite de Conexiones"
            Login_Fail(Usuario)
        End If

    End Sub



    Private Sub Validar(ByVal Usuario As String, ByVal Password As String)
        Try

            Zamba.Users.Factory.RightFactory.ValidateLogIn(Usuario, Password)

            If Zamba.Users.Factory.RightFactory.CurrentUser Is Nothing Then
                Login_Fail(Usuario)
            End If

            Me.MyUser = Zamba.Users.Factory.RightFactory.CurrentUser

            If Zamba.Users.Factory.RightFactory.CurrentUser.ID <> -1 AndAlso Zamba.Users.Factory.RightFactory.CurrentUser.ID <> 0 Then
                If SecurityOptions.ClaveVencida(Zamba.Users.Factory.RightFactory.CurrentUser.ID) = False Then
                    Login_Ok(Usuario)
                Else
                    Me.Panel1.Visible = False
                    Me.Panel2.Visible = True
                End If
            Else
                If Zamba.Users.Factory.RightFactory.CurrentUser.ID = -1 Then
                    Me.FailureText.Text = "El Usuario se encuentra bloqueado"
                Else
                    Me.FailureText.Text = "El Usuario o la clave ingresadas son incorréctos"
                End If
                Login_Fail(Usuario)
            End If


        Catch ex As System.Threading.ThreadAbortException
        Catch ex As Exception
            Me.FailureText.Text = ex.Message
            Login_Fail(Usuario)
        End Try

    End Sub



    Public ReadOnly Property Usuario() As String
        Get
            Return Me.UserName.Text
        End Get
    End Property



    Protected Sub LoginButton_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles LoginButton.Click
        If Me.UserName.Text = "" OrElse Me.Password.Text = "" Then Exit Sub
        Validar(Me.UserName.Text, Me.Password.Text)
	End Sub



    Protected Sub Button1_Click(ByVal sender As Object, ByVal e As System.EventArgs) Handles Button1.Click

        If Me.TextBox2.Text = Me.TextBox1.Text Then

            If SecurityOptions.IsValidPassword(Me.MyUser.Id, Usuario.Trim, Me.TextBox1.Text.Trim, True) Then
                Me.MyUser.Password = Me.TextBox1.Text.Trim
                Try
                    Zamba.Users.Factory.UserFactory.Update(Me.MyUser)
                    'Esta dentro del user.Update Zamba.Security.ZSecurityOptions.SecurityOptions.SavePassword(Me._usr.id, Me.Login_Password.Text)
                    'MessageBox.Show("La clave se cambió correctamente", "Zamba Software", MessageBoxButtons.OK, MessageBoxIcon.Information)
                    'RaiseEvent ChangePassword(True)
                    Me.Panel2.Visible = False
                    Me.Panel1.Visible = True
                Catch ex As Exception
                    RaiseError(ex)
                    Login_Fail(Usuario)
                    'MessageBox.Show("Ocurrio un error al intentar cambiar la clave. " & ex.ToString, "Zamba Software", MessageBoxButtons.OK, MessageBoxIcon.Error)
                End Try

            End If
        End If

    End Sub


End Class
